In [1]:
import numpy as np
import pandas as pd

In [2]:
books = pd.read_csv(r"C:\Users\Parikshit\Desktop\Coding\Data science\AI-ML for internship\Projects for internships\Book Recommender system\Dataset\Books.csv")
users = pd.read_csv(r"C:\Users\Parikshit\Desktop\Coding\Data science\AI-ML for internship\Projects for internships\Book Recommender system\Dataset\Users.csv")
ratings = pd.read_csv(r"C:\Users\Parikshit\Desktop\Coding\Data science\AI-ML for internship\Projects for internships\Book Recommender system\Dataset\Ratings.csv")

C:\Users\Parikshit\AppData\Local\Temp\ipykernel_14972\4052046920.py:1: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  books = pd.read_csv(r"C:\Users\Parikshit\Desktop\Coding\Data science\AI-ML for internship\Projects for internships\Book Recommender system\Dataset\Books.csv")


In [3]:
books.head()

ISBN                                         Book-Title  \
0  0195153448                                Classical Mythology   
1  0002005018                                       Clara Callan   
2  0060973129                               Decision in Normandy   
3  0374157065  Flu: The Story of the Great Influenza Pandemic...   
4  0393045218                             The Mummies of Urumchi   

            Book-Author Year-Of-Publication                   Publisher  \
0    Mark P. O. Morford                2002     Oxford University Press   
1  Richard Bruce Wright                2001       HarperFlamingo Canada   
2          Carlo D'Este                1991             HarperPerennial   
3      Gina Bari Kolata                1999        Farrar Straus Giroux   
4       E. J. W. Barber                1999  W. W. Norton &amp; Company   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-M  \
0  http://images.amazon.com/images/P/0195153448.0...   
1  http://images.amazon.com/images/P/0002005018.0...   
2  http://images.amazon.com/images/P/0060973129.0...   
3  http://images.amazon.com/images/P/0374157065.0...   
4  http://images.amazon.com/images/P/0393045218.0...   

                                         Image-URL-L  
0  http://images.amazon.com/images/P/0195153448.0...  
1  http://images.amazon.com/images/P/0002005018.0...  
2  http://images.amazon.com/images/P/0060973129.0...  
3  http://images.amazon.com/images/P/0374157065.0...  
4  http://images.amazon.com/images/P/0393045218.0...

In [4]:
users.head()

User-ID                            Location   Age
0        1                  nyc, new york, usa   NaN
1        2           stockton, california, usa  18.0
2        3     moscow, yukon territory, russia   NaN
3        4           porto, v.n.gaia, portugal  17.0
4        5  farnborough, hants, united kingdom   NaN

In [5]:
ratings.head()

User-ID        ISBN  Book-Rating
0   276725  034545104X            0
1   276726  0155061224            5
2   276727  0446520802            0
3   276729  052165615X            3
4   276729  0521795028            6

In [6]:
books.isnull().sum()

ISBN                   0
Book-Title             0
Book-Author            2
Year-Of-Publication    0
Publisher              2
Image-URL-S            0
Image-URL-M            0
Image-URL-L            3
dtype: int64

In [7]:
users.isnull().sum()

User-ID          0
Location         0
Age         110762
dtype: int64

In [8]:
ratings.isnull().sum()

User-ID        0
ISBN           0
Book-Rating    0
dtype: int64

In [9]:
books.duplicated().sum()

0

In [10]:
users.duplicated().sum()

0

In [11]:
ratings.duplicated().sum()

0

### EDA

In [12]:
books.describe()

ISBN      Book-Title      Book-Author  Year-Of-Publication  \
count       271360          271360           271358               271360   
unique      271360          242135           102022                  202   
top     0195153448  Selected Poems  Agatha Christie                 2002   
freq             1              27              632                13903   

        Publisher                                        Image-URL-S  \
count      271358                                             271360   
unique      16807                                             271044   
top     Harlequin  http://images.amazon.com/images/P/185326119X.0...   
freq         7535                                                  2   

                                              Image-URL-M  \
count                                              271360   
unique                                             271044   
top     http://images.amazon.com/images/P/185326119X.0...   
freq                                                    2   

                                              Image-URL-L  
count                                              271357  
unique                                             271041  
top     http://images.amazon.com/images/P/225307649X.0...  
freq                                                    2

In [13]:
users.describe()

User-ID            Age
count  278858.00000  168096.000000
mean   139429.50000      34.751434
std     80499.51502      14.428097
min         1.00000       0.000000
25%     69715.25000      24.000000
50%    139429.50000      32.000000
75%    209143.75000      44.000000
max    278858.00000     244.000000

In [14]:
ratings.describe()

User-ID   Book-Rating
count  1.149780e+06  1.149780e+06
mean   1.403864e+05  2.866950e+00
std    8.056228e+04  3.854184e+00
min    2.000000e+00  0.000000e+00
25%    7.034500e+04  0.000000e+00
50%    1.410100e+05  0.000000e+00
75%    2.110280e+05  7.000000e+00
max    2.788540e+05  1.000000e+01

In [15]:
# merging ratings and books dataset
ratings_with_name = ratings.merge(books,on='ISBN')

In [16]:
ratings_with_name

User-ID        ISBN  Book-Rating  \
0         276725  034545104X            0   
1           2313  034545104X            5   
2           6543  034545104X            0   
3           8680  034545104X            5   
4          10314  034545104X            9   
...          ...         ...          ...   
1031131   276688  0517145553            0   
1031132   276688  1575660792            7   
1031133   276690  0590907301            0   
1031134   276704  0679752714            0   
1031135   276704  0806917695            5   

                                                Book-Title      Book-Author  \
0                                     Flesh Tones: A Novel       M. J. Rose   
1                                     Flesh Tones: A Novel       M. J. Rose   
2                                     Flesh Tones: A Novel       M. J. Rose   
3                                     Flesh Tones: A Novel       M. J. Rose   
4                                     Flesh Tones: A Novel       M. J. Rose   
...                                                    ...              ...   
1031131                                    Mostly Harmless    Douglas Adams   
1031132                                        Gray Matter  Shirley Kennett   
1031133  Triplet Trouble and the Class Trip (Triplet Tr...     Debbie Dadey   
1031134  A Desert of Pure Feeling (Vintage Contemporaries)   Judith Freeman   
1031135  Perplexing Lateral Thinking Puzzles: Scholasti...      Paul Sloane   

        Year-Of-Publication                          Publisher  \
0                      2002                   Ballantine Books   
1                      2002                   Ballantine Books   
2                      2002                   Ballantine Books   
3                      2002                   Ballantine Books   
4                      2002                   Ballantine Books   
...                     ...                                ...   
1031131                1995             Random House Value Pub   
1031132                1996  Kensington Publishing Corporation   
1031133                1997                              Apple   
1031134                1997                  Vintage Books USA   
1031135                1997                Sterling Publishing   

                                               Image-URL-S  \
0        http://images.amazon.com/images/P/034545104X.0...   
1        http://images.amazon.com/images/P/034545104X.0...   
2        http://images.amazon.com/images/P/034545104X.0...   
3        http://images.amazon.com/images/P/034545104X.0...   
4        http://images.amazon.com/images/P/034545104X.0...   
...                                                    ...   
1031131  http://images.amazon.com/images/P/0517145553.0...   
1031132  http://images.amazon.com/images/P/1575660792.0...   
1031133  http://images.amazon.com/images/P/0590907301.0...   
1031134  http://images.amazon.com/images/P/0679752714.0...   
1031135  http://images.amazon.com/images/P/0806917695.0...   

                                               Image-URL-M  \
0        http://images.amazon.com/images/P/034545104X.0...   
1        http://images.amazon.com/images/P/034545104X.0...   
2        http://images.amazon.com/images/P/034545104X.0...   
3        http://images.amazon.com/images/P/034545104X.0...   
4        http://images.amazon.com/images/P/034545104X.0...   
...                                                    ...   
1031131  http://images.amazon.com/images/P/0517145553.0...   
1031132  http://images.amazon.com/images/P/1575660792.0...   
1031133  http://images.amazon.com/images/P/0590907301.0...   
1031134  http://images.amazon.com/images/P/0679752714.0...   
1031135  http://images.amazon.com/images/P/0806917695.0...   

                                               Image-URL-L  
0        http://images.amazon.com/images/P/034545104X.0...  
1        http://images.amazon.com/images/P/034545104X.0...  
2        http://images.amazon.com/images/P/034545104X

In [17]:
ratings_with_name.groupby('Book-Title').count()['Book-Rating'].reset_index()

Book-Title  Book-Rating
0        A Light in the Storm: The Civil War Diary of ...            4
1                                   Always Have Popsicles            1
2                    Apple Magic (The Collector's series)            1
3        Ask Lily (Young Women of Faith: Lily Series, ...            1
4        Beyond IBM: Leadership Marketing and Finance ...            1
...                                                   ...          ...
241066                                      Ã?Â?lpiraten.            2
241067                     Ã?Â?rger mit Produkt X. Roman.            4
241068                                Ã?Â?sterlich leben.            1
241069                              Ã?Â?stlich der Berge.            3
241070                                  Ã?Â?thique en toc            2

[241071 rows x 2 columns]

In [18]:
book_ratings_df = ratings_with_name.groupby('Book-Title').count()['Book-Rating'].reset_index()
book_ratings_df.rename(columns={'Book-Rating':'num_ratings'},inplace=True)
book_ratings_df.sample(10)

Book-Title  num_ratings
24903      Blake's Selected Poems (Dover Thrift Editions)            8
50580       Die Entdeckung des Alphabets. Erz�¤hlungen.            1
166421                                       Star Witness            6
151475                                 Romiette and Julio            4
51104   Die Nacht, die kein Ende nahm. In der Gewalt v...            1
145938                                   Quitting the Mob            1
186248  The End of the Affair (Twentieth Century Class...           17
164539               Space Exploration (Eyewitness Books)            1
16812              Ascension: John Coltrane and His Quest            1
39113   Constructing a life philosophy: An examination...            1

In [19]:
# Convert 'Book-Rating' to string to identify non-numeric values
ratings_with_name['Book-Rating'] = ratings_with_name['Book-Rating'].astype(str)

# Remove rows with non-numeric values
ratings_with_name = ratings_with_name[ratings_with_name['Book-Rating'].str.isnumeric()]

# Convert 'Book-Rating' to numeric
ratings_with_name['Book-Rating'] = pd.to_numeric(ratings_with_name['Book-Rating'], errors='coerce')

# Drop rows with NaN values in 'Book-Rating'
ratings_with_name.dropna(subset=['Book-Rating'], inplace=True)

# Convert 'Book-Rating' to integer
ratings_with_name['Book-Rating'] = ratings_with_name['Book-Rating'].astype(int)

In [20]:
avg_ratings_df = ratings_with_name.groupby('Book-Title').apply(lambda x: np.mean(x['Book-Rating'])).reset_index(name='avg_ratings')
avg_ratings_df.rename(columns={'Book-Rating':'avg_ratings'},inplace=True)
avg_ratings_df

Book-Title  avg_ratings
0        A Light in the Storm: The Civil War Diary of ...     2.250000
1                                   Always Have Popsicles     0.000000
2                    Apple Magic (The Collector's series)     0.000000
3        Ask Lily (Young Women of Faith: Lily Series, ...     8.000000
4        Beyond IBM: Leadership Marketing and Finance ...     0.000000
...                                                   ...          ...
241066                                      Ã?Â?lpiraten.     0.000000
241067                     Ã?Â?rger mit Produkt X. Roman.     5.250000
241068                                Ã?Â?sterlich leben.     7.000000
241069                              Ã?Â?stlich der Berge.     2.666667
241070                                  Ã?Â?thique en toc     4.000000

[241071 rows x 2 columns]

In [21]:
avg_ratings_df['avg_ratings'] = avg_ratings_df['avg_ratings'].round(2)
avg_ratings_df

Book-Title  avg_ratings
0        A Light in the Storm: The Civil War Diary of ...         2.25
1                                   Always Have Popsicles         0.00
2                    Apple Magic (The Collector's series)         0.00
3        Ask Lily (Young Women of Faith: Lily Series, ...         8.00
4        Beyond IBM: Leadership Marketing and Finance ...         0.00
...                                                   ...          ...
241066                                      Ã?Â?lpiraten.         0.00
241067                     Ã?Â?rger mit Produkt X. Roman.         5.25
241068                                Ã?Â?sterlich leben.         7.00
241069                              Ã?Â?stlich der Berge.         2.67
241070                                  Ã?Â?thique en toc         4.00

[241071 rows x 2 columns]

In [22]:
popular_df = book_ratings_df.merge(avg_ratings_df,on='Book-Title')
popular_df

Book-Title  num_ratings  \
0        A Light in the Storm: The Civil War Diary of ...            4   
1                                   Always Have Popsicles            1   
2                    Apple Magic (The Collector's series)            1   
3        Ask Lily (Young Women of Faith: Lily Series, ...            1   
4        Beyond IBM: Leadership Marketing and Finance ...            1   
...                                                   ...          ...   
241066                                      Ã?Â?lpiraten.            2   
241067                     Ã?Â?rger mit Produkt X. Roman.            4   
241068                                Ã?Â?sterlich leben.            1   
241069                              Ã?Â?stlich der Berge.            3   
241070                                  Ã?Â?thique en toc            2   

        avg_ratings  
0              2.25  
1              0.00  
2              0.00  
3              8.00  
4              0.00  
...             ...  
241066         0.00  
241067         5.25  
241068         7.00  
241069         2.67  
241070         4.00  

[241071 rows x 3 columns]

In [23]:
popular_df[popular_df['num_ratings']>=250]

Book-Title  num_ratings  \
764                                                  1984          284   
818                                   1st to Die: A Novel          509   
1048                                           2nd Chance          356   
1760                                   A Bend in the Road          346   
2281    A Child Called \It\": One Child's Courage to S...          265   
...                                                   ...          ...   
233850                           White Oleander : A Novel          387   
233851       White Oleander : A Novel (Oprah's Book Club)          356   
234740  Wicked: The Life and Times of the Wicked Witch...          326   
234951                                        Wild Animus         2502   
240860                                \O\" Is for Outlaw"          269   

        avg_ratings  
764            4.45  
818            3.58  
1048           3.27  
1760           3.36  
2281           4.09  
...             ...  
233850         3.51  
233851         3.77  
234740         3.77  
234951         1.02  
240860         2.99  

[186 rows x 3 columns]

In [24]:
popular_df[popular_df['num_ratings']>=250].sort_values('avg_ratings',ascending=False).head(50)

Book-Title  num_ratings  \
80434   Harry Potter and the Prisoner of Azkaban (Book 3)          428   
80422        Harry Potter and the Goblet of Fire (Book 4)          387   
80441      Harry Potter and the Sorcerer's Stone (Book 1)          278   
80426   Harry Potter and the Order of the Phoenix (Boo...          347   
80414    Harry Potter and the Chamber of Secrets (Book 2)          556   
191612  The Hobbit : The Enchanting Prelude to The Lor...          281   
187377  The Fellowship of the Ring (The Lord of the Ri...          368   
80445   Harry Potter and the Sorcerer's Stone (Harry P...          575   
211384     The Two Towers (The Lord of the Rings, Part 2)          260   
219741                              To Kill a Mockingbird          510   
183573                                  The Da Vinci Code          898   
187880                 The Five People You Meet in Heaven          430   
180556                             The Catcher in the Rye          449   
196326                          The Lovely Bones: A Novel         1295   
206502                            The Secret Life of Bees          774   
144165                           Prodigal Summer: A Novel          253   
128670                                         Neverwhere          265   
764                                                  1984          284   
168719  Stupid White Men ...and Other Sorry Excuses fo...          283   
223135  Tuesdays with Morrie: An Old Man, a Young Man,...          493   
204387                The Red Tent (Bestselling Backlist)          723   
191589               The Hitchhiker's Guide to the Galaxy          268   
129379   Nickel and Dimed: On (Not) Getting By in America          335   
93381                                       Into the Wild          252   
63867                                      Fahrenheit 451          409   
74750                           Girl with a Pearl Earring          526   
136145                                          Outlander          283   
233370  Where the Heart Is (Oprah's Book Club (Paperba...          585   
156102                     Seabiscuit: An American Legend          275   
107962                                         Life of Pi          664   
176845                                     The Bean Trees          389   
2281    A Child Called \It\": One Child's Courage to S...          265   
8434                                       ANGELA'S ASHES          279   
76343                                         Good in Bed          490   
64931   Fast Food Nation: The Dark Side of the All-Ame...          321   
12700                                       American Gods          302   
161645                                 Skipping Christmas          322   
189551    The Golden Compass (His Dark Materials, Book 1)          336   
105777  Left Behind: A Novel of the Earth's Last Days ...          318   
181679                                   The Color Purple          314   
160336                               Silence of the Lambs          256   
8752                                          About a Boy          262   
158138                  Seven Up (A Stephanie Plum Novel)          278   
175097  The Alchemist: A Fable About Following Your Dream          266   
80069   Hard Eight : A Stephanie Plum Novel (A Stephan...          269   
170101                       Suzanne's Diary for Nicholas          457   
111073                                  Lord of the Flies          259   
5664                              A Prayer for Owen Meany          413   
212070   The Vampire Lestat (Vampire Chronicles, Book II)          301   
234740  Wicked: The Life and Times of the Wicked Witch...          326   

        avg_ratings  
80434          5.85  
80422          5.82  
80441          5.74  
80426          5.50  
80414          5.18  
191612         5.01  
187377         4.95  
80445          4.90  
211384         4.88  
219741         4.70  
183573         4.64  
187880  

In [25]:
popular_df = popular_df[popular_df['num_ratings']>=250].sort_values('avg_ratings',ascending=False).head(50)

In [26]:
popular_df.merge(books,on='Book-Title')

Book-Title  num_ratings  \
0    Harry Potter and the Prisoner of Azkaban (Book 3)          428   
1    Harry Potter and the Prisoner of Azkaban (Book 3)          428   
2    Harry Potter and the Prisoner of Azkaban (Book 3)          428   
3         Harry Potter and the Goblet of Fire (Book 4)          387   
4         Harry Potter and the Goblet of Fire (Book 4)          387   
..                                                 ...          ...   
191                                  Lord of the Flies          259   
192                            A Prayer for Owen Meany          413   
193                            A Prayer for Owen Meany          413   
194   The Vampire Lestat (Vampire Chronicles, Book II)          301   
195  Wicked: The Life and Times of the Wicked Witch...          326   

     avg_ratings        ISBN         Book-Author Year-Of-Publication  \
0           5.85  0439136350       J. K. Rowling                1999   
1           5.85  0439136369       J. K. Rowling                2001   
2           5.85  0786222743       J. K. Rowling                2000   
3           5.82  0439139597       J. K. Rowling                2000   
4           5.82  0439139600       J. K. Rowling                2002   
..           ...         ...                 ...                 ...   
191         3.82  0698102193  William G. Golding                1978   
192         3.80  0345361792         John Irving                1990   
193         3.80  0345363523         John Irving                1989   
194         3.78  0345313860           ANNE RICE                1986   
195         3.77  0060987103     Gregory Maguire                1996   

                 Publisher                                        Image-URL-S  \
0               Scholastic  http://images.amazon.com/images/P/0439136350.0...   
1               Scholastic  http://images.amazon.com/images/P/0439136369.0...   
2          Thorndike Press  http://images.amazon.com/images/P/0786222743.0...   
3               Scholastic  http://images.amazon.com/images/P/0439139597.0...   
4    Scholastic Paperbacks  http://images.amazon.com/images/P/0439139600.0...   
..                     ...                                                ...   
191       Putnam Pub Group  http://images.amazon.com/images/P/0698102193.0...   
192       Ballantine Books  http://images.amazon.com/images/P/0345361792.0...   
193       Ballantine Books  http://images.amazon.com/images/P/0345363523.0...   
194       Ballantine Books  http://images.amazon.com/images/P/0345313860.0...   
195            Regan Books  http://images.amazon.com/images/P/0060987103.0...   

                                           Image-URL-M  \
0    http://images.amazon.com/images/P/0439136350.0...   
1    http://images.amazon.com/images/P/0439136369.0...   
2    http://images.amazon.com/images/P/0786222743.0...   
3    http://images.amazon.com/images/P/0439139597.0...   
4    http://images.amazon.com/images/P/0439139600.0...   
..                                                 ...   
191  http://images.amazon.com/images/P/0698102193.0...   
192  http://images.amazon.com/images/P/0345361792.0...   
193  http://images.amazon.com/images/P/0345363523.0...   
194  http://images.amazon.com/images/P/0345313860.0...   
195  http://images.amazon.com/images/P/0060987103.0...   

                                           Image-URL-L  
0    http://images.amazon.com/images/P/0439136350.0...  
1    http://images.amazon.com/images/P/0439136369.0...  
2    http://images.amazon.com/images/P/0786222743.0...  
3    http://images.amazon.com/images/P/0439139597.0...  
4    http://images.amazon.com/images/P/0439139600.0...  
..                                                 ...  
191  http://images.amazon.com/images/P/0698102193.0...  
192  http://images.amazon.com/images/P/0345361792.0...  
193  http://images.amazon.com/images/P/0345363523.0...  
194  http://images.amazon.com/images/P/0345313860.0...  
195  http://images.amazon.com

In [27]:
 popular_df.merge(books,on='Book-Title').drop_duplicates('Book-Title')

Book-Title  num_ratings  \
0    Harry Potter and the Prisoner of Azkaban (Book 3)          428   
3         Harry Potter and the Goblet of Fire (Book 4)          387   
5       Harry Potter and the Sorcerer's Stone (Book 1)          278   
9    Harry Potter and the Order of the Phoenix (Boo...          347   
13    Harry Potter and the Chamber of Secrets (Book 2)          556   
16   The Hobbit : The Enchanting Prelude to The Lor...          281   
17   The Fellowship of the Ring (The Lord of the Ri...          368   
26   Harry Potter and the Sorcerer's Stone (Harry P...          575   
28      The Two Towers (The Lord of the Rings, Part 2)          260   
39                               To Kill a Mockingbird          510   
47                                   The Da Vinci Code          898   
53                  The Five People You Meet in Heaven          430   
55                              The Catcher in the Rye          449   
62                           The Lovely Bones: A Novel         1295   
63                             The Secret Life of Bees          774   
69                            Prodigal Summer: A Novel          253   
70                                          Neverwhere          265   
75                                                1984          284   
84   Stupid White Men ...and Other Sorry Excuses fo...          283   
85   Tuesdays with Morrie: An Old Man, a Young Man,...          493   
88                 The Red Tent (Bestselling Backlist)          723   
89                The Hitchhiker's Guide to the Galaxy          268   
98    Nickel and Dimed: On (Not) Getting By in America          335   
100                                      Into the Wild          252   
103                                     Fahrenheit 451          409   
116                          Girl with a Pearl Earring          526   
117                                          Outlander          283   
123  Where the Heart Is (Oprah's Book Club (Paperba...          585   
124                     Seabiscuit: An American Legend          275   
126                                         Life of Pi          664   
130                                     The Bean Trees          389   
133  A Child Called \It\": One Child's Courage to S...          265   
135                                     ANGELA'S ASHES          279   
137                                        Good in Bed          490   
139  Fast Food Nation: The Dark Side of the All-Ame...          321   
140                                      American Gods          302   
141                                 Skipping Christmas          322   
147    The Golden Compass (His Dark Materials, Book 1)          336   
155  Left Behind: A Novel of the Earth's Last Days ...          318   
157                                   The Color Purple          314   
163                               Silence of the Lambs          256   
166                                        About a Boy          262   
173                  Seven Up (A Stephanie Plum Novel)          278   
175  The Alchemist: A Fable About Following Your Dream          266   
177  Hard Eight : A Stephanie Plum Novel (A Stephan...          269   
182                       Suzanne's Diary for Nicholas          457   
186                                  Lord of the Flies          259   
192                            A Prayer for Owen Meany          413   
194   The Vampire Lestat (Vampire Chronicles, Book II)          301   
195  Wicked: The Life and Times of the Wicked Witch...          326   

     avg_ratings        ISBN             Book-Author Year-Of-Publication  \
0           5.85  0439136350           J. K. Rowling                1999   
3           5.82  0439139597           J. K. Rowling                2000   
5           5.74  0590353403           J. K. Rowling                1998   
9           5.50  043935806X           J. K. Rowling                2003   
13          5.18  0439064872           J. 

In [28]:
 popular_df.merge(books,on='Book-Title').drop_duplicates('Book-Title').shape

(50, 10)

In [29]:
 popular_df.merge(books,on='Book-Title').drop_duplicates('Book-Title')[['Book-Title','Book-Author','Image-URL-M','num_ratings','avg_ratings']]

Book-Title  \
0    Harry Potter and the Prisoner of Azkaban (Book 3)   
3         Harry Potter and the Goblet of Fire (Book 4)   
5       Harry Potter and the Sorcerer's Stone (Book 1)   
9    Harry Potter and the Order of the Phoenix (Boo...   
13    Harry Potter and the Chamber of Secrets (Book 2)   
16   The Hobbit : The Enchanting Prelude to The Lor...   
17   The Fellowship of the Ring (The Lord of the Ri...   
26   Harry Potter and the Sorcerer's Stone (Harry P...   
28      The Two Towers (The Lord of the Rings, Part 2)   
39                               To Kill a Mockingbird   
47                                   The Da Vinci Code   
53                  The Five People You Meet in Heaven   
55                              The Catcher in the Rye   
62                           The Lovely Bones: A Novel   
63                             The Secret Life of Bees   
69                            Prodigal Summer: A Novel   
70                                          Neverwhere   
75                                                1984   
84   Stupid White Men ...and Other Sorry Excuses fo...   
85   Tuesdays with Morrie: An Old Man, a Young Man,...   
88                 The Red Tent (Bestselling Backlist)   
89                The Hitchhiker's Guide to the Galaxy   
98    Nickel and Dimed: On (Not) Getting By in America   
100                                      Into the Wild   
103                                     Fahrenheit 451   
116                          Girl with a Pearl Earring   
117                                          Outlander   
123  Where the Heart Is (Oprah's Book Club (Paperba...   
124                     Seabiscuit: An American Legend   
126                                         Life of Pi   
130                                     The Bean Trees   
133  A Child Called \It\": One Child's Courage to S...   
135                                     ANGELA'S ASHES   
137                                        Good in Bed   
139  Fast Food Nation: The Dark Side of the All-Ame...   
140                                      American Gods   
141                                 Skipping Christmas   
147    The Golden Compass (His Dark Materials, Book 1)   
155  Left Behind: A Novel of the Earth's Last Days ...   
157                                   The Color Purple   
163                               Silence of the Lambs   
166                                        About a Boy   
173                  Seven Up (A Stephanie Plum Novel)   
175  The Alchemist: A Fable About Following Your Dream   
177  Hard Eight : A Stephanie Plum Novel (A Stephan...   
182                       Suzanne's Diary for Nicholas   
186                                  Lord of the Flies   
192                            A Prayer for Owen Meany   
194   The Vampire Lestat (Vampire Chronicles, Book II)   
195  Wicked: The Life and Times of the Wicked Witch...   

                Book-Author  \
0             J. K. Rowling   
3             J. K. Rowling   
5             J. K. Rowling   
9             J. K. Rowling   
13            J. K. Rowling   
16           J.R.R. TOLKIEN   
17           J.R.R. TOLKIEN   
26            J. K. Rowling   
28           J.R.R. TOLKIEN   
39               Harper Lee   
47                Dan Brown   
53              Mitch Albom   
55            J.D. Salinger   
62             Alice Sebold   
63            Sue Monk Kidd   
69       Barbara Kingsolver   
70              Neil Gaiman   
75            George Orwell   
84            Michael Moore   
85              MITCH ALBOM   
88            Anita Diamant   
89            Douglas Adams   
98       Barbara Ehrenreich   
100            Jon Krakauer   
103            Ray Bradbury   
116         Tracy Chevalier   
117          DIANA GABALDON   
123            Billie Letts   
124       LAURA HILLENBRAND   
126             Yann Martel   
130      Barbara Kingsolver   
133             Dave Pelzer   
135           Frank McCourt   
137         Jennifer Weiner   


In [30]:
popular_df = popular_df.merge(books,on='Book-Title').drop_duplicates('Book-Title')[['Book-Title','Book-Author','Image-URL-M','num_ratings','avg_ratings']]

## Collaborative filtering 

In [31]:
ratings_with_name.groupby('User-ID').count()

ISBN  Book-Rating  Book-Title  Book-Author  Year-Of-Publication  \
User-ID                                                                    
2           1            1           1            1                    1   
8          17           17          17           17                   17   
9           3            3           3            3                    3   
10          1            1           1            1                    1   
12          1            1           1            1                    1   
...       ...          ...         ...          ...                  ...   
278846      1            1           1            1                    1   
278849      4            4           4            4                    4   
278851     23           23          23           23                   23   
278852      1            1           1            1                    1   
278854      8            8           8            8                    8   

         Publisher  Image-URL-S  Image-URL-M  Image-URL-L  
User-ID                                                    
2                1            1            1            1  
8               17           17           17           17  
9                3            3            3            3  
10               1            1            1            1  
12               1            1            1            1  
...            ...          ...          ...          ...  
278846           1            1            1            1  
278849           4            4            4            4  
278851          23           23           23           23  
278852           1            1            1            1  
278854           8            8            8            8  

[92106 rows x 9 columns]

In [32]:
x = ratings_with_name.groupby('User-ID').count()['Book-Rating'] > 200
knowledgeable_users = x[x].index

In [33]:
knowledgeable_users

Index([   254,   2276,   2766,   2977,   3363,   4017,   4385,   6251,   6323,
         6543,
       ...
       271705, 273979, 274004, 274061, 274301, 274308, 275970, 277427, 277639,
       278418],
      dtype='int64', name='User-ID', length=811)

In [34]:
# these are the ratings that are being done by the knowledgeable_users 
filtered_rating = ratings_with_name[ratings_with_name['User-ID'].isin(knowledgeable_users)]

In [35]:
y = filtered_rating.groupby('Book-Title').count()['Book-Rating']>=50
famous_books = y[y].index

In [36]:
#final ratings of those books whose ratings are above 50
final_ratings = filtered_rating[filtered_rating['Book-Title'].isin(famous_books)]
final_ratings

User-ID        ISBN  Book-Rating              Book-Title  \
63        278418  0446520802            0            The Notebook   
65          3363  0446520802            0            The Notebook   
66          7158  0446520802           10            The Notebook   
69         11676  0446520802           10            The Notebook   
74         23768  0446520802            6            The Notebook   
...          ...         ...          ...                     ...   
1026724   266865  0531001725           10  The Catcher in the Rye   
1027923   269566  0670809381            0                  Echoes   
1028777   271284  0440910927            0           The Rainmaker   
1029070   271705  B0001PIOX4            0          Fahrenheit 451   
1030868   275970  1586210661            9  Me Talk Pretty One Day   

                   Book-Author Year-Of-Publication                Publisher  \
63             Nicholas Sparks                1996             Warner Books   
65             Nicholas Sparks                1996             Warner Books   
66             Nicholas Sparks                1996             Warner Books   
69             Nicholas Sparks                1996             Warner Books   
74             Nicholas Sparks                1996             Warner Books   
...                        ...                 ...                      ...   
1026724  Jerome David Salinger                1973   Scholastic Library Pub   
1027923           Maeve Binchy                1986              Penguin USA   
1028777           John Grisham                1995                   Island   
1029070           Ray Bradbury                1993     Simon &amp; Schuster   
1030868          David Sedaris                2001  Time Warner Audio Major   

                                               Image-URL-S  \
63       http://images.amazon.com/images/P/0446520802.0...   
65       http://images.amazon.com/images/P/0446520802.0...   
66       http://images.amazon.com/images/P/0446520802.0...   
69       http://images.amazon.com/images/P/0446520802.0...   
74       http://images.amazon.com/images/P/0446520802.0...   
...                                                    ...   
1026724  http://images.amazon.com/images/P/0531001725.0...   
1027923  http://images.amazon.com/images/P/0670809381.0...   
1028777  http://images.amazon.com/images/P/0440910927.0...   
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...   
1030868  http://images.amazon.com/images/P/1586210661.0...   

                                               Image-URL-M  \
63       http://images.amazon.com/images/P/0446520802.0...   
65       http://images.amazon.com/images/P/0446520802.0...   
66       http://images.amazon.com/images/P/0446520802.0...   
69       http://images.amazon.com/images/P/0446520802.0...   
74       http://images.amazon.com/images/P/0446520802.0...   
...                                                    ...   
1026724  http://images.amazon.com/images/P/0531001725.0...   
1027923  http://images.amazon.com/images/P/0670809381.0...   
1028777  http://images.amazon.com/images/P/0440910927.0...   
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...   
1030868  http://images.amazon.com/images/P/1586210661.0...   

                                               Image-URL-L  
63       http://images.amazon.com/images/P/0446520802.0...  
65       http://images.amazon.com/images/P/0446520802.0...  
66       http://images.amazon.com/images/P/0446520802.0...  
69       http://images.amazon.com/images/P/0446520802.0...  
74       http://images.amazon.com/images/P/0446520802.0...  
...                                                    ...  
1026724  http://images.amazon.com/images/P/0531001725.0...  
1027923  http://images.amazon.com/images/P/0670809381.0...  
1028777  http://images.amazon.com/images/P/0440910927.0...  
1029070  http://images.amazon.com/images/P/B0001PIOX4.0...  
1030868  http://images.amazon.com/images/P/1586210661.0...  

[58586 r

In [37]:
final_ratings.pivot_table(index='Book-Title',columns='User-ID',values='Book-Rating')

User-ID                                             254     2276    2766    \
Book-Title                                                                   
1984                                                   9.0     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN    10.0     NaN   
4 Blondes                                              NaN     NaN     NaN   
A Bend in the Road                                     0.0     NaN     7.0   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     NaN     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             2977    3363    4017    \
Book-Title                                                                   
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     NaN     NaN   
A Bend in the Road                                     NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        7.0     NaN     NaN   
You Belong To Me                                       NaN     NaN     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     0.0     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             4385    6251    6323    \
Book-Title                                                                   
1984                                                   NaN     NaN     NaN   
1st to Die: A Novel                                    NaN     NaN     NaN   
2nd Chance                                             NaN     NaN     NaN   
4 Blondes                                              NaN     0.0     NaN   
A Bend in the Road                                     NaN     NaN     NaN   
...                                                    ...     ...     ...   
Year of Wonders                                        NaN     NaN     NaN   
You Belong To Me                                       NaN     NaN     0.0   
Zen and the Art of Motorcycle Maintenance: An I...     NaN     0.0     NaN   
Zoya                                                   NaN     NaN     NaN   
\O\" Is for Outlaw"                                    NaN     NaN     NaN   

User-ID                                             6543    ...  271705  \
Book-Title                                                  ...           
1984                                                   NaN  ...    10.0   
1st to Die: A Novel                                    9.0  ...     NaN   
2nd Chance                                             0.0  ...     NaN   
4 Blondes                                              NaN  ...     NaN   
A Bend in the Road                                     NaN  ...     NaN   
...                                                    ...  ...     ...   
Year of Wonders                                        0.0  ...     NaN   
You Belong To Me                                       NaN  ...     NaN   
Zen and the Art of Motorcycle Maintenance: An I...     NaN  ...     NaN   
Zoya                                                   NaN  ...     NaN   
\O\" Is for Outlaw"                                    

The format of data we need for our filtering

In [38]:
final_pt = final_ratings.pivot_table(index='Book-Title',columns='User-ID',values='Book-Rating')

In [39]:
final_pt.fillna(0,inplace=True)
final_pt.head(5)

User-ID              254     2276    2766    2977    3363    4017    4385    \
Book-Title                                                                    
1984                    9.0     0.0     0.0     0.0     0.0     0.0     0.0   
1st to Die: A Novel     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
2nd Chance              0.0    10.0     0.0     0.0     0.0     0.0     0.0   
4 Blondes               0.0     0.0     0.0     0.0     0.0     0.0     0.0   
A Bend in the Road      0.0     0.0     7.0     0.0     0.0     0.0     0.0   

User-ID              6251    6323    6543    ...  271705  273979  274004  \
Book-Title                                   ...                           
1984                    0.0     0.0     0.0  ...    10.0     0.0     0.0   
1st to Die: A Novel     0.0     0.0     9.0  ...     0.0     0.0     0.0   
2nd Chance              0.0     0.0     0.0  ...     0.0     0.0     0.0   
4 Blondes               0.0     0.0     0.0  ...     0.0     0.0     0.0   
A Bend in the Road      0.0     0.0     0.0  ...     0.0     0.0     0.0   

User-ID              274061  274301  274308  275970  277427  277639  278418  
Book-Title                                                                   
1984                    0.0     0.0     0.0     0.0     0.0     0.0     0.0  
1st to Die: A Novel     0.0     0.0     0.0     0.0     0.0     0.0     0.0  
2nd Chance              0.0     0.0     0.0     0.0     0.0     0.0     0.0  
4 Blondes               0.0     0.0     0.0     0.0     0.0     0.0     0.0  
A Bend in the Road      0.0     0.0     0.0     0.0     0.0     0.0     0.0  

[5 rows x 810 columns]

Here every row is a point in an 810 dimensional plane

In [40]:
# to find the similiarity between different data points we do the following
from sklearn.metrics.pairwise import cosine_similarity
cosine_similarity(final_pt)

array([[1.        , 0.10255025, 0.01220856, ..., 0.12110367, 0.07347567,
        0.04316046],
       [0.10255025, 1.        , 0.2364573 , ..., 0.07446129, 0.16773875,
        0.14263397],
       [0.01220856, 0.2364573 , 1.        , ..., 0.04558758, 0.04938579,
        0.10796119],
       ...,
       [0.12110367, 0.07446129, 0.04558758, ..., 1.        , 0.07085128,
        0.0196177 ],
       [0.07347567, 0.16773875, 0.04938579, ..., 0.07085128, 1.        ,
        0.10602962],
       [0.04316046, 0.14263397, 0.10796119, ..., 0.0196177 , 0.10602962,
        1.        ]])

In [41]:
cosine_similarity(final_pt).shape

(706, 706)

euclidean distance of 706 books with 706 books in the multi-dimensional plane

In [42]:
similarity_scores = cosine_similarity(final_pt)

In [47]:
def recommend(book_name):
    #index fetch
    index = np.where(final_pt.index==book_name)[0][0]
    similar_items = sorted(list(enumerate(similarity_scores[index])),key=lambda x:x[1],reverse=True)[1:6]

    data = []
    for i in similar_items:
        item = []
        temp_df = books[books['Book-Title'] == final_pt.index[i[0]]]
        item.extend(list(temp_df.drop_duplicates('Book-Title')['Book-Title'].values))
        item.extend(list(temp_df.drop_duplicates('Book-Title')['Book-Author'].values))
        item.extend(list(temp_df.drop_duplicates('Book-Title')['Image-URL-M'].values))
        
        data.append(item)
    
    return data

In [48]:
recommend('1984')

[['Animal Farm',
  'George Orwell',
  'http://images.amazon.com/images/P/0451526341.01.MZZZZZZZ.jpg'],
 ["The Handmaid's Tale",
  'Margaret Atwood',
  'http://images.amazon.com/images/P/0449212602.01.MZZZZZZZ.jpg'],
 ['Brave New World',
  'Aldous Huxley',
  'http://images.amazon.com/images/P/0060809833.01.MZZZZZZZ.jpg'],
 ['The Vampire Lestat (Vampire Chronicles, Book II)',
  'ANNE RICE',
  'http://images.amazon.com/images/P/0345313860.01.MZZZZZZZ.jpg'],
 ['The Hours : A Novel',
  'Michael Cunningham',
  'http://images.amazon.com/images/P/0312243022.01.MZZZZZZZ.jpg']]

In [44]:
import pickle
pickle.dump(popular_df, open('popular.pkl','wb'))

In [49]:
pickle.dump(final_pt,open('pt.pkl','wb'))
pickle.dump(books,open('books.pkl','wb'))
pickle.dump(similarity_scores,open('similarity_scores.pkl','wb'))